pypeit/io.py

```diff
@@ -851,40 +851,48 @@ def create_symlink(filename, symlink_dir, relative_symlink=False, overwrite=Fals
     """
-    Grab the list of files with a given extension 
+    Find files from one or more paths with one or more extensions.
 
-    Args:
-        raw_path (str or list):
-            Path(s) to raw files, which may or may not include the prefix of the
-            files to search for.  
-
-            For a string input, for example, this can be the directory
-            ``'/path/to/files/'`` or the directory plus the file prefix
-            ``'/path/to/files/prefix'``, which yeilds the search strings
-            ``'/path/to/files/*fits'`` or ``'/path/to/files/prefix*fits'``,
-            respectively.
-
-            For a list input, this can use wildcards for multiple directories.
+    This is a recursive function.  If ``raw_path`` is a list, the function is
+    called for every item in the list and the results are concatenated.
 
-        extension (str, optional):
-            File extension to search on.
+    Args:
+        raw_path (:obj:`str`, `Path`_, :obj:`list`):
+            One or more paths to search for files, which may or may not include
+            the prefix of the files to search for.  For string input, this can
+            be the directory ``'/path/to/files/'`` or the directory plus the
+            file prefix ``'/path/to/files/prefix'``, which yeilds the search
+            strings ``'/path/to/files/*fits'`` or
+            ``'/path/to/files/prefix*fits'``, respectively.  For a list input,
+            this can use wildcards for multiple directories.
+
+        extension (:obj:`str`, :obj:`list`, optional):
+            One or more file extensions to search on.
 
     Returns:
-        list: List of raw data filenames (sorted) with full path
-    """
-    if isinstance(raw_path, str):
-        # Grab the list of files
-        dfname = os.path.join(raw_path, f'*{extension}*') \
-                    if os.path.isdir(raw_path) else f'{raw_path}*{extension}*'
-        return sorted(glob.glob(dfname))
+        :obj:`list`: List of `Path`_ objects with the full path to the set of
+        unique raw data filenames that match the provided criteria search
+        strings.
+    """
+    if isinstance(raw_path, (str, Path)):
+        _raw_path = Path(raw_path).absolute()
+        if _raw_path.is_dir():
+            prefix = ''
+        else:
+            _raw_path, prefix = _raw_path.parent, _raw_path.name
+            if not _raw_path.is_dir():
+                msgs.error(f'{_raw_path} does not exist!')
+        ext = [extension] if isinstance(extension, str) else extension
+        files = numpy.concatenate([sorted(_raw_path.glob(f'{prefix}*{e}')) for e in ext])
+        return numpy.unique(files).tolist()
     
     if isinstance(raw_path, list):
-        return numpy.concatenate([files_from_extension(p, extension=extension) for p in raw_path]).tolist()
+        files = numpy.concatenate([files_from_extension(p, extension=extension) for p in raw_path])
+        return numpy.unique(files).tolist()
```

Comment:
I think we possibly need to check if there are strings present in this list that are identical, except for the extension? I would be doubtful that anyone would carry around identical compressed and uncompressed files in the same directory, but then again, it's possible.

Reply:
I'd rather not over-complicate it.  I think it's enough to warn people that the code will potentially find both compressed and uncompressed files if they're in the same directory, which has always been true.

Reply:
OK - sounds good to me! 👍